compaction: use better partition estimate for split compaction

lkshminarayanan opened the pull request:
Split compaction divides the partitions in an existing sstable into two groups and writes them into two new sstables, which replace the original one. The partition count from the original sstable is used as an estimate when writing the new ones, but this estimate is not accurate as the partitions are split between the two new sstables and each will contain only a portion of the original partition count. This also causes the bloom filters to be rebuilt at the end of compaction, as they were initially built with inaccurate estimates.

Fix this by using a better estimate for the output sstables, which is half the original partition count.

Fixes #20253

Improvement; No need to backport.

avikivity commented on compaction/compaction.cc:
```diff
@@ -1364,6 +1364,12 @@ class cleanup_compaction final : public regular_compaction {
+        // one sstable will be split into two, so use half the
+        // total estimated partitions for the output tables.
+        return _estimated_partitions / 2;
+    }
```
A better estimate is a linear interpolation based on the token range spanned by the sstable. If the sstable spans the token range (1000, 2000) and we split at 1100, the left-hand-side would have 10% of the partitions and the right-hand side 90%.

I see partitions_per_sstable() can't give such a response, but maybe it can be adjusted.

denesb replied:
We can use `sstables::sstable::estimated_keys_for_range()`, it estimates partition-count for a given token-range, based on the number of summary entries that fall into said range. It should be better than linear interpolation.

lkshminarayanan replied:
I have updated the PR to come up with a better partition estimate.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12065/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12065/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/895/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3953/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11970/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12065/)
#### Build Details:
- Duration: 7 hr 3 min
- Builder: i-0762a18cade65369d (m5ad.8xlarge)

lkshminarayanan commented:
Changelog V2 :
 - Calculate the resultant ranges after a split and estimate the partitions based on those ranges instead of doing 1/2 of the total estimate.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12431/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12431/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/compaction_group_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12431/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4218/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12241/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12431/)
#### Build Details:
- Duration: 3 hr 55 min
- Builder: spider7.cloudius-systems.com

lkshminarayanan commented:
> ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4218/)
> ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12241/)

Pipeline failure. I'll retrigger after sometime.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12473/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12473/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/compaction_group_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12473/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12274/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1183/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4250/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12473/)
#### Build Details:
- Duration: 7 hr 20 min
- Builder: spider8.cloudius-systems.com

raphaelsc commented on locator/tablets.cc:
```diff
@@ -310,6 +310,19 @@ dht::token_range tablet_map::get_token_range(tablet_id id) const {
+    auto get_last_token_post_split = [&] (tablet_id id) {
+        return dht::last_token_of_compaction_group(_log2_tablets + 1, size_t(id));
+    };
+
```
let's introduce a 

get_last_token(tablet_id id, log2_tablets).

get_last_token(id) calls get_last_token(id, _log2_tablets)

then we use the former here.

lkshminarayanan replied:
Done. I added couple of variants - `get_last_token_impl` and `get_token_range_impl` that helped with the final implementation.

raphaelsc commented on replica/database.hh:
```diff
@@ -579,6 +579,8 @@ public:
     // be split once it returns.
     future<> maybe_split_compaction_group_of(locator::tablet_id);
+
+    dht::token_range get_post_split_token_range_of(const dht::token&) const noexcept;
```
I think it requires some doc since it's not clear what it is about.

lkshminarayanan replied:
Done.

raphaelsc commented on compaction/compaction.cc:
```diff
@@ -1386,6 +1386,20 @@ class split_compaction final : public regular_compaction {
+
+        auto token_range_of_new_sst = _table_s.get_post_split_token_range_of(dk.token());
+        const auto estimated_keys = _sstables[0]->estimated_keys_for_range(token_range_of_new_sst);
+
```
Nice. did you test it manually? A simple test is to populate a tablet table and wait for splits. --target-tablet-size-in-bytes 20971520 shorten the time for splits to happen.

lkshminarayanan replied:
I tested them in parts as I was not able to successfully trigger a split. I'll use your method and test it before submitting the next version.

lkshminarayanan replied:
By reducing the `target-tablet-size-in-bytes`, I was able to trigger multiple splits on an sstable with 50K partitions. 

Without this fix, the split was invoked seven times before tablets were finally balanced and bloom filter was rebuilt twice for every split (14 rebuilds).
With the fix, the rebuild came down to 7 times in my attempts for the same number of splits.

Log of a test from when the sstable was flushed until the tablets were balanced : 

Note : There are multiple WARN messages in the log that emit the `estimated keys` value calculated by the patch - they are there because I was trying to get insight into the calculation and they are not a part of the final patch. 

```
INFO  2024-10-22 20:57:38,075 [shard 0:strm] api - perform_keyspace_flush: keyspace=ks tables=["test"]
INFO  2024-10-22 20:58:26,436 [shard 0: gms] raft_topology - raft topology: Refreshing table load stats for DC datacenter1 that has 1 token owners
INFO  2024-10-22 20:58:26,438 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 20:58:26,438 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=1, tablets=1, shards=1, state=normal
INFO  2024-10-22 20:58:26,438 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 20:58:26,438 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=1 has an average tablet size of 2311227
INFO  2024-10-22 20:58:26,438 [shard 0: gms] load_balancer - Emitting resize decision of type split for table 2984a460-908a-11ef-93f4-09ea64b31b3e due to avg tablet size of 2311227
INFO  2024-10-22 20:58:26,438 [shard 0: gms] load_balancer - Prepared 1 migration plans, out of which there were 0 tablet migration(s) and 1 resize decision(s)
INFO  2024-10-22 20:58:26,438 [shard 0: gms] raft_topology - updating topology state: Starting tablet migration
INFO  2024-10-22 20:58:26,457 [shard 0: gms] raft_topology - entered `tablet migration` transition state
INFO  2024-10-22 20:58:26,457 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 20:58:26,457 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=1, tablets=1, shards=1, state=normal
INFO  2024-10-22 20:58:26,457 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 20:58:26,457 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=1 has an average tablet size of 2311227
INFO  2024-10-22 20:58:26,457 [shard 0: gms] load_balancer - Prepared 0 migration plans, out of which there were 0 tablet migration(s) and 0 resize decision(s)
INFO  2024-10-22 20:58:26,457 [shard 0: gms] raft_topology - updating topology state: Finished tablet migration
INFO  2024-10-22 20:58:26,457 [shard 0:strm] compaction - [Split ks.test 484b8491-908a-11ef-93f4-09ea64b31b3e] Splitting [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_16y2_0gi4w28zy0sfxww666-big-Data.db:level=0:origin=memtable]
WARN  2024-10-22 20:58:26,458 [shard 0:strm] compaction - [Split ks.test 484b8491-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_16ze_2q5y828zy0sfxww666-big-Data.db` : 25024
INFO  2024-10-22 20:58:26,466 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 20:58:26,466 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=1, tablets=1, shards=1, state=normal
INFO  2024-10-22 20:58:26,466 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 20:58:26,466 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=1 has an average tablet size of 2311227
INFO  2024-10-22 20:58:26,466 [shard 0: gms] load_balancer - Prepared 0 migration plans, out of which there were 0 tablet migration(s) and 0 resize decision(s)
INFO  2024-10-22 20:58:26,591 [shard 0:strm] sstable - Rebuilding bloom filter /Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_16ze_2q5y828zy0sfxww666-big-Filter.db: resizing bitset from 31288 bytes to 31336 bytes. sstable origin: split
WARN  2024-10-22 20:58:26,614 [shard 0:strm] compaction - [Split ks.test 484b8491-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_16ze_3nlnk28zy0sfxww666-big-Data.db` : 25831
INFO  2024-10-22 20:58:26,761 [shard 0:strm] compaction - [Split ks.test 484b8491-908a-11ef-93f4-09ea64b31b3e] Split 1 sstables to [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_16ze_2q5y828zy0sfxww666-big-Data.db:level=0,/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_16ze_3nlnk28zy0sfxww666-big-Data.db:level=0]. 2MB to 2MB (~100% of original) in 298ms = 7MB/s. ~50048 total partitions merged to 49999.
INFO  2024-10-22 20:58:26,762 [shard 0:strm] table - Setting split ready sequence number to 1 for table ks.test
INFO  2024-10-22 20:59:26,437 [shard 0: gms] raft_topology - raft topology: Refreshing table load stats for DC datacenter1 that has 1 token owners
INFO  2024-10-22 20:59:26,438 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 20:59:26,438 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=1, tablets=1, shards=1, state=normal
INFO  2024-10-22 20:59:26,438 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 20:59:26,438 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=1 has an average tablet size of 2311657
INFO  2024-10-22 20:59:26,438 [shard 0: gms] load_balancer - Finalizing resize decision for table 2984a460-908a-11ef-93f4-09ea64b31b3e as all replicas agree on sequence number 1
INFO  2024-10-22 20:59:26,438 [shard 0: gms] load_balancer - Prepared 1 migration plans, out of which there were 0 tablet migration(s) and 1 resize decision(s)
INFO  2024-10-22 20:59:26,438 [shard 0: gms] raft_topology - updating topology state: Started tablet split finalization
INFO  2024-10-22 20:59:26,456 [shard 0: gms] raft_topology - entered `tablet split finalization` transition state
INFO  2024-10-22 20:59:26,456 [shard 0: gms] raft_topology - executing global topology command barrier_and_drain, excluded nodes: {}
INFO  2024-10-22 20:59:26,456 [shard 0: gms] raft_topology - updating topology state: advance fence version to 6
INFO  2024-10-22 20:59:26,463 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 20:59:26,463 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=1, tablets=1, shards=1, state=normal
INFO  2024-10-22 20:59:26,463 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 20:59:26,463 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=1 has an average tablet size of 2311657
INFO  2024-10-22 20:59:26,463 [shard 0: gms] load_balancer - Finalizing resize decision for table 2984a460-908a-11ef-93f4-09ea64b31b3e as all replicas agree on sequence number 1
INFO  2024-10-22 20:59:26,463 [shard 0: gms] load_balancer - Prepared 1 migration plans, out of which there were 0 tablet migration(s) and 1 resize decision(s)
INFO  2024-10-22 20:59:26,463 [shard 0: gms] load_balancer - Split tablets for table 2984a460-908a-11ef-93f4-09ea64b31b3e, increasing tablet count from 1 to 2
INFO  2024-10-22 20:59:26,463 [shard 0: gms] raft_topology - updating topology state: Finished tablet split finalization
INFO  2024-10-22 20:59:26,469 [shard 0: gms] table - Detected tablet split for table ks.test, increasing from 1 to 2 tablets
INFO  2024-10-22 20:59:26,470 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 20:59:26,471 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=2, tablets=2, shards=1, state=normal
INFO  2024-10-22 20:59:26,471 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 20:59:26,471 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=2 has an average tablet size of 1155828
INFO  2024-10-22 20:59:26,471 [shard 0: gms] load_balancer - Emitting resize decision of type split for table 2984a460-908a-11ef-93f4-09ea64b31b3e due to avg tablet size of 1155828
INFO  2024-10-22 20:59:26,471 [shard 0: gms] load_balancer - Prepared 1 migration plans, out of which there were 0 tablet migration(s) and 1 resize decision(s)
INFO  2024-10-22 20:59:26,471 [shard 0: gms] raft_topology - updating topology state: Starting tablet migration
INFO  2024-10-22 20:59:26,478 [shard 0: gms] raft_topology - entered `tablet migration` transition state
INFO  2024-10-22 20:59:26,478 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 20:59:26,478 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=2, tablets=2, shards=1, state=normal
INFO  2024-10-22 20:59:26,479 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 20:59:26,479 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=2 has an average tablet size of 1155828
INFO  2024-10-22 20:59:26,479 [shard 0: gms] load_balancer - Prepared 0 migration plans, out of which there were 0 tablet migration(s) and 0 resize decision(s)
INFO  2024-10-22 20:59:26,479 [shard 0: gms] raft_topology - updating topology state: Finished tablet migration
INFO  2024-10-22 20:59:26,479 [shard 0:strm] compaction - [Split ks.test 6c11fee1-908a-11ef-93f4-09ea64b31b3e] Splitting [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_16ze_3nlnk28zy0sfxww666-big-Data.db:level=0:origin=split]
WARN  2024-10-22 20:59:26,479 [shard 0:strm] compaction - [Split ks.test 6c11fee1-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_2unzk28zy0sfxww666-big-Data.db` : 12882
INFO  2024-10-22 20:59:26,487 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 20:59:26,487 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=2, tablets=2, shards=1, state=normal
INFO  2024-10-22 20:59:26,487 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 20:59:26,487 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=2 has an average tablet size of 1155828
INFO  2024-10-22 20:59:26,487 [shard 0: gms] load_balancer - Prepared 0 migration plans, out of which there were 0 tablet migration(s) and 0 resize decision(s)
WARN  2024-10-22 20:59:26,559 [shard 0:strm] compaction - [Split ks.test 6c11fee1-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_3bt9s28zy0sfxww666-big-Data.db` : 12882
INFO  2024-10-22 20:59:26,677 [shard 0:strm] compaction - [Split ks.test 6c11fee1-908a-11ef-93f4-09ea64b31b3e] Split 1 sstables to [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_2unzk28zy0sfxww666-big-Data.db:level=0,/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_3bt9s28zy0sfxww666-big-Data.db:level=0]. 1MB to 1MB (~100% of original) in 194ms = 5MB/s. ~24960 total partitions merged to 24937.
INFO  2024-10-22 20:59:26,678 [shard 0:strm] compaction - [Split ks.test 6c305c51-908a-11ef-93f4-09ea64b31b3e] Splitting [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_16ze_2q5y828zy0sfxww666-big-Data.db:level=0:origin=split]
WARN  2024-10-22 20:59:26,678 [shard 0:strm] compaction - [Split ks.test 6c305c51-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_41bhc28zy0sfxww666-big-Data.db` : 12948
WARN  2024-10-22 20:59:26,780 [shard 0:strm] compaction - [Split ks.test 6c305c51-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_4n6io28zy0sfxww666-big-Data.db` : 12948
INFO  2024-10-22 20:59:26,876 [shard 0:strm] compaction - [Split ks.test 6c305c51-908a-11ef-93f4-09ea64b31b3e] Split 1 sstables to [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_41bhc28zy0sfxww666-big-Data.db:level=0,/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_4n6io28zy0sfxww666-big-Data.db:level=0]. 1MB to 1MB (~100% of original) in 195ms = 5MB/s. ~25088 total partitions merged to 25062.
INFO  2024-10-22 20:59:26,877 [shard 0:strm] table - Setting split ready sequence number to 3 for table ks.test
INFO  2024-10-22 21:00:26,438 [shard 0: gms] raft_topology - raft topology: Refreshing table load stats for DC datacenter1 that has 1 token owners
INFO  2024-10-22 21:00:26,439 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 21:00:26,439 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=2, tablets=2, shards=1, state=normal
INFO  2024-10-22 21:00:26,439 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 21:00:26,439 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=2 has an average tablet size of 1161101
INFO  2024-10-22 21:00:26,439 [shard 0: gms] load_balancer - Finalizing resize decision for table 2984a460-908a-11ef-93f4-09ea64b31b3e as all replicas agree on sequence number 3
INFO  2024-10-22 21:00:26,439 [shard 0: gms] load_balancer - Prepared 1 migration plans, out of which there were 0 tablet migration(s) and 1 resize decision(s)
INFO  2024-10-22 21:00:26,439 [shard 0: gms] raft_topology - updating topology state: Started tablet split finalization
INFO  2024-10-22 21:00:26,455 [shard 0: gms] raft_topology - entered `tablet split finalization` transition state
INFO  2024-10-22 21:00:26,456 [shard 0: gms] raft_topology - executing global topology command barrier_and_drain, excluded nodes: {}
INFO  2024-10-22 21:00:26,456 [shard 0: gms] raft_topology - updating topology state: advance fence version to 10
INFO  2024-10-22 21:00:26,464 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 21:00:26,464 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=2, tablets=2, shards=1, state=normal
INFO  2024-10-22 21:00:26,464 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 21:00:26,464 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=2 has an average tablet size of 1161101
INFO  2024-10-22 21:00:26,464 [shard 0: gms] load_balancer - Finalizing resize decision for table 2984a460-908a-11ef-93f4-09ea64b31b3e as all replicas agree on sequence number 3
INFO  2024-10-22 21:00:26,464 [shard 0: gms] load_balancer - Prepared 1 migration plans, out of which there were 0 tablet migration(s) and 1 resize decision(s)
INFO  2024-10-22 21:00:26,464 [shard 0: gms] load_balancer - Split tablets for table 2984a460-908a-11ef-93f4-09ea64b31b3e, increasing tablet count from 2 to 4
INFO  2024-10-22 21:00:26,464 [shard 0: gms] raft_topology - updating topology state: Finished tablet split finalization
INFO  2024-10-22 21:00:26,470 [shard 0: gms] table - Detected tablet split for table ks.test, increasing from 2 to 4 tablets
INFO  2024-10-22 21:00:26,472 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 21:00:26,472 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=4, tablets=4, shards=1, state=normal
INFO  2024-10-22 21:00:26,472 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 21:00:26,472 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=4 has an average tablet size of 580550
INFO  2024-10-22 21:00:26,472 [shard 0: gms] load_balancer - Emitting resize decision of type split for table 2984a460-908a-11ef-93f4-09ea64b31b3e due to avg tablet size of 580550
INFO  2024-10-22 21:00:26,472 [shard 0: gms] load_balancer - Prepared 1 migration plans, out of which there were 0 tablet migration(s) and 1 resize decision(s)
INFO  2024-10-22 21:00:26,472 [shard 0: gms] raft_topology - updating topology state: Starting tablet migration
INFO  2024-10-22 21:00:26,479 [shard 0: gms] raft_topology - entered `tablet migration` transition state
INFO  2024-10-22 21:00:26,479 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 21:00:26,479 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=4, tablets=4, shards=1, state=normal
INFO  2024-10-22 21:00:26,479 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 21:00:26,479 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=4 has an average tablet size of 580550
INFO  2024-10-22 21:00:26,479 [shard 0: gms] load_balancer - Prepared 0 migration plans, out of which there were 0 tablet migration(s) and 0 resize decision(s)
INFO  2024-10-22 21:00:26,479 [shard 0: gms] raft_topology - updating topology state: Finished tablet migration
INFO  2024-10-22 21:00:26,480 [shard 0:strm] compaction - [Split ks.test 8fd56bf1-908a-11ef-93f4-09ea64b31b3e] Splitting [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_41bhc28zy0sfxww666-big-Data.db:level=0:origin=split]
WARN  2024-10-22 21:00:26,480 [shard 0:strm] compaction - [Split ks.test 8fd56bf1-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_2uvpc28zy0sfxww666-big-Data.db` : 6208
INFO  2024-10-22 21:00:26,488 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 21:00:26,488 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=4, tablets=4, shards=1, state=normal
INFO  2024-10-22 21:00:26,489 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 21:00:26,489 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=4 has an average tablet size of 580550
INFO  2024-10-22 21:00:26,489 [shard 0: gms] load_balancer - Prepared 0 migration plans, out of which there were 0 tablet migration(s) and 0 resize decision(s)
INFO  2024-10-22 21:00:26,526 [shard 0:strm] sstable - Rebuilding bloom filter /Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_2uvpc28zy0sfxww666-big-Filter.db: resizing bitset from 7768 bytes to 7856 bytes. sstable origin: split
WARN  2024-10-22 21:00:26,536 [shard 0:strm] compaction - [Split ks.test 8fd56bf1-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_36vsw28zy0sfxww666-big-Data.db` : 6984
INFO  2024-10-22 21:00:26,574 [shard 0:strm] sstable - Rebuilding bloom filter /Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_36vsw28zy0sfxww666-big-Filter.db: resizing bitset from 8736 bytes to 7600 bytes. sstable origin: split
INFO  2024-10-22 21:00:26,587 [shard 0:strm] compaction - [Split ks.test 8fd56bf1-908a-11ef-93f4-09ea64b31b3e] Split 1 sstables to [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_2uvpc28zy0sfxww666-big-Data.db:level=0,/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_36vsw28zy0sfxww666-big-Data.db:level=0]. 573kB to 577kB (~100% of original) in 104ms = 5MB/s. ~12416 total partitions merged to 12350.
INFO  2024-10-22 21:00:26,587 [shard 0:strm] compaction - [Split ks.test 8fe5e6b1-908a-11ef-93f4-09ea64b31b3e] Splitting [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_3bt9s28zy0sfxww666-big-Data.db:level=0:origin=split]
WARN  2024-10-22 21:00:26,588 [shard 0:strm] compaction - [Split ks.test 8fe5e6b1-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_3i11c28zy0sfxww666-big-Data.db` : 6208
WARN  2024-10-22 21:00:26,631 [shard 0:strm] compaction - [Split ks.test 8fe5e6b1-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_3r8ts28zy0sfxww666-big-Data.db` : 6984
INFO  2024-10-22 21:00:26,671 [shard 0:strm] sstable - Rebuilding bloom filter /Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_3r8ts28zy0sfxww666-big-Filter.db: resizing bitset from 8736 bytes to 7848 bytes. sstable origin: split
INFO  2024-10-22 21:00:26,686 [shard 0:strm] compaction - [Split ks.test 8fe5e6b1-908a-11ef-93f4-09ea64b31b3e] Split 1 sstables to [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_3i11c28zy0sfxww666-big-Data.db:level=0,/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_3r8ts28zy0sfxww666-big-Data.db:level=0]. 576kB to 580kB (~100% of original) in 95ms = 6MB/s. ~12416 total partitions merged to 12402.
INFO  2024-10-22 21:00:26,687 [shard 0:strm] compaction - [Split ks.test 8ff528f1-908a-11ef-93f4-09ea64b31b3e] Splitting [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_2unzk28zy0sfxww666-big-Data.db:level=0:origin=split]
WARN  2024-10-22 21:00:26,687 [shard 0:strm] compaction - [Split ks.test 8ff528f1-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_438xe28zy0sfxww666-big-Data.db` : 6272
WARN  2024-10-22 21:00:26,734 [shard 0:strm] compaction - [Split ks.test 8ff528f1-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_4dbkw28zy0sfxww666-big-Data.db` : 7056
INFO  2024-10-22 21:00:26,771 [shard 0:strm] sstable - Rebuilding bloom filter /Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_4dbkw28zy0sfxww666-big-Filter.db: resizing bitset from 8824 bytes to 7960 bytes. sstable origin: split
INFO  2024-10-22 21:00:26,787 [shard 0:strm] compaction - [Split ks.test 8ff528f1-908a-11ef-93f4-09ea64b31b3e] Split 1 sstables to [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_438xe28zy0sfxww666-big-Data.db:level=0,/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_4dbkw28zy0sfxww666-big-Data.db:level=0]. 582kB to 586kB (~100% of original) in 97ms = 6MB/s. ~12544 total partitions merged to 12535.
INFO  2024-10-22 21:00:26,787 [shard 0:strm] compaction - [Split ks.test 90046b31-908a-11ef-93f4-09ea64b31b3e] Splitting [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_4n6io28zy0sfxww666-big-Data.db:level=0:origin=split]
WARN  2024-10-22 21:00:26,788 [shard 0:strm] compaction - [Split ks.test 90046b31-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_4ow8w28zy0sfxww666-big-Data.db` : 6400
INFO  2024-10-22 21:00:26,826 [shard 0:strm] sstable - Rebuilding bloom filter /Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_4ow8w28zy0sfxww666-big-Filter.db: resizing bitset from 8008 bytes to 8024 bytes. sstable origin: split
WARN  2024-10-22 21:00:26,837 [shard 0:strm] compaction - [Split ks.test 90046b31-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_4zec028zy0sfxww666-big-Data.db` : 7200
INFO  2024-10-22 21:00:26,877 [shard 0:strm] sstable - Rebuilding bloom filter /Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_4zec028zy0sfxww666-big-Filter.db: resizing bitset from 9008 bytes to 7880 bytes. sstable origin: split
INFO  2024-10-22 21:00:26,890 [shard 0:strm] compaction - [Split ks.test 90046b31-908a-11ef-93f4-09ea64b31b3e] Split 1 sstables to [/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_4ow8w28zy0sfxww666-big-Data.db:level=0,/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_172q_4zec028zy0sfxww666-big-Data.db:level=0]. 590kB to 594kB (~100% of original) in 99ms = 5MB/s. ~12800 total partitions merged to 12712.
INFO  2024-10-22 21:00:26,891 [shard 0:strm] table - Setting split ready sequence number to 5 for table ks.test
INFO  2024-10-22 21:01:26,438 [shard 0: gms] raft_topology - raft topology: Refreshing table load stats for DC datacenter1 that has 1 token owners
INFO  2024-10-22 21:01:26,439 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 21:01:26,439 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=4, tablets=4, shards=1, state=normal
INFO  2024-10-22 21:01:26,439 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 21:01:26,439 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=4 has an average tablet size of 585003
INFO  2024-10-22 21:01:26,439 [shard 0: gms] load_balancer - Finalizing resize decision for table 2984a460-908a-11ef-93f4-09ea64b31b3e as all replicas agree on sequence number 5
INFO  2024-10-22 21:01:26,439 [shard 0: gms] load_balancer - Prepared 1 migration plans, out of which there were 0 tablet migration(s) and 1 resize decision(s)
INFO  2024-10-22 21:01:26,439 [shard 0: gms] raft_topology - updating topology state: Started tablet split finalization
INFO  2024-10-22 21:01:26,456 [shard 0: gms] raft_topology - entered `tablet split finalization` transition state
INFO  2024-10-22 21:01:26,456 [shard 0: gms] raft_topology - executing global topology command barrier_and_drain, excluded nodes: {}
INFO  2024-10-22 21:01:26,456 [shard 0: gms] raft_topology - updating topology state: advance fence version to 14
INFO  2024-10-22 21:01:26,464 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 21:01:26,464 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=4, tablets=4, shards=1, state=normal
INFO  2024-10-22 21:01:26,464 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
INFO  2024-10-22 21:01:26,464 [shard 0: gms] load_balancer - Table 2984a460-908a-11ef-93f4-09ea64b31b3e with tablet_count=4 has an average tablet size of 585003
INFO  2024-10-22 21:01:26,464 [shard 0: gms] load_balancer - Finalizing resize decision for table 2984a460-908a-11ef-93f4-09ea64b31b3e as all replicas agree on sequence number 5
INFO  2024-10-22 21:01:26,464 [shard 0: gms] load_balancer - Prepared 1 migration plans, out of which there were 0 tablet migration(s) and 1 resize decision(s)
INFO  2024-10-22 21:01:26,464 [shard 0: gms] load_balancer - Split tablets for table 2984a460-908a-11ef-93f4-09ea64b31b3e, increasing tablet count from 4 to 8
INFO  2024-10-22 21:01:26,465 [shard 0: gms] raft_topology - updating topology state: Finished tablet split finalization
INFO  2024-10-22 21:01:26,471 [shard 0: gms] table - Detected tablet split for table ks.test, increasing from 4 to 8 tablets
INFO  2024-10-22 21:01:26,472 [shard 0: gms] load_balancer - Examining DC datacenter1 (shuffle=false, balancing=true)
INFO  2024-10-22 21:01:26,472 [shard 0: gms] load_balancer - Node 52c54c31-1c54-493f-a9c9-6e0dc962a6a9: rack=rack1 avg_load=8, tablets=8, shards=1, state=normal
INFO  2024-10-22 21:01:26,472 [shard 0: gms] load_balancer - Prepared 0 migrations in DC datacenter1
```

lkshminarayanan replied:
The patch didn't help if the partitions were too large. I tried with few large partitions, each having rows in the range of [10, 100000] and with the patch, after a split most of the rebuilds were still happening.

raphaelsc replied:
> - [Split ks.test 6c305c51-908a-11ef-93f4-09ea64b31b3e] estimated keys for `/Scylla/workdir/data/ks/test-2984a460908a11ef93f409ea64b31b3e/me-3gkl_1712_4n6io28zy0sfxww666-big-Data.db` : 12948

I can think of index sampling level (summary) having an impact, but with large partitions, I'd expect each large partition to have its own summary entry to due to data:summary byte ratio of 2000:1.

lkshminarayanan replied:
One possible reason: 

The logic that decides if a filter has to be rebuilt, skips rebuilding if the false positive rate is within +/-25% of the configured fp rate. For small Bloom filters, even a minor variation in the estimate can easily push the false positive rate beyond the allowed limit, triggering rebuilds. 

I tested with 100 large partitions, and for 100 keys, the bloom filter size is already small : 128 bytes. When split, the bloom filter sizes keep decreasing.  Maybe that is why I'm seeing many rebuilds.

lkshminarayanan commented:
Changelog V2 :
 - Refactored the new tablet_map methods to reduce code duplication.

raphaelsc commented on locator/tablets.cc:
```diff
@@ -299,14 +303,26 @@ dht::token tablet_map::get_first_token(tablet_id id) const {
+}
+
+dht::token_range tablet_map::get_token_range_after_split(const token& t) const noexcept {
+    // when the tablets are split, the tablet count doubles, (i.e.) _log2_tablets increases by 1
```
bonus point for the new name :-)

raphaelsc commented on locator/tablets.hh:
```diff
@@ -327,6 +327,13 @@ private:
+    /// Returns token_range which contains all tokens owned by the specified tablet
+    /// when the tablet_count is `1 << log2_tablets`.
+    dht::token_range get_token_range_impl(tablet_id id, size_t log2_tablets) const;
 public:
```
I think it would be better to name them get_last_token() and get_token_range(), since they have different param set than public ones. But shouldn't block this series.

lkshminarayanan replied:
Done.

lkshminarayanan commented:
> I think it would be better to name them get_last_token() and get_token_range(), since they have different param set than public ones. But shouldn't block this series.

Changelog V3 :
 - rebased on top of latest master
 - renamed the methods as per above suggestion.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13091/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13091/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/compaction_group_test
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/171/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1634/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12737/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4709/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13091/)
#### Build Details:
- Duration: 4 hr 1 min
- Builder: spider1.cloudius-systems.com
